### ISSUE TITLE
"ScraploiAuthenticationError" on C886 with IOS 15.7(3)M4a and no Login attempt at all

### ISSUE THREAD

**Issue by 991jo** [carlmontanari/scrapli#38]
Hello

I just made a little test with a Cisco C886 running IOS (not IOS-XE) Version 15.7(3)M4a.
I am not sure how the support for IOS devices looks in scrapli. officially only IOS-XE is supported but from a cli perspective I have not yet found a fundamental difference in the CLI so I would assume that it just works with the IOS-XE driver.

Here is my example Code:

```
from scrapli.driver.core import IOSXEDriver

r1 = {  "host": "10.0.0.1",
        "auth_username": "test",
        "auth_password": "test",
        "auth_strict_key": False,
        "transport_options": {"open_cmd": ["-o", "KexAlgorithms=+diffie-hellman-group-exchange-sha1", "-o", "Ciphers+=aes256-cbc"]},
        "ssh_config_file": False
        }

router = IOSXEDriver(**r1)
router.open()
```

This fails with the following stacktrace:
```
Failed to open connection to host 10.0.0.1. Do you need to disable `auth_strict_key`?
Traceback (most recent call last):
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/transport/ptyprocess.py", line 369, in read
    s = self.fileobj.read1(size)
OSError: [Errno 5] Input/output error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/transport/systemssh.py", line 369, in _authenticate
    new_output = self.session.read()
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/transport/ptyprocess.py", line 374, in read
    raise EOFError("End Of File (EOF). Exception style platform.")
EOFError: End Of File (EOF). Exception style platform.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 12, in <module>
    router.open()
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/driver/driver.py", line 92, in open
    self.transport.open()
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/transport/systemssh.py", line 256, in open
    self._open_pty()
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/transport/systemssh.py", line 334, in _open_pty
    self._authenticate()
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/decorators.py", line 72, in timeout_wrapper
    result = future.get(timeout=timeout_duration)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 771, in get
    raise self._value
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/jo/projects/scrapli-tests/lib/python3.8/site-packages/scrapli/transport/systemssh.py", line 382, in _authenticate
    raise ScrapliAuthenticationFailed(msg)
scrapli.exceptions.ScrapliAuthenticationFailed: Failed to open connection to host 10.0.0.1. Do you need to disable `auth_strict_key`?
```
the device is reachable and I can log into the device with the appropriate ssh settings.
changing the value of `auth_strict_key` does not change the behaviour.
When I am running tcpdump it does not show a connection attempt at all.

Am I missing something or is plain IOS just not supported at all?
Let me know if you need any futher information.

**Comment by carlmontanari** [comment]
Yeah you are right, while not "officially" supported IOS should for sure work. I've tested this against an old 2960 I have at home and other than having to fixup the older ssh cipher stuff as you've already done it has worked as expected.

Can you get two pieces of info for me so we can try to troubleshoot this:

1. output from manually ssh'ing to the device in a terminal (wondering if there is a prompt pattern or something that we are not accounting for)
2. logs of scrapli attempting to connect. [here](https://github.com/carlmontanari/scrapli/blob/master/examples/logging/basic_logging.py) is a simple example to setup logging, this should get us all the data scrapli reads and give us a good idea of what's up

Basically what happens in the system ssh auth is that we read from the terminal until we cant read anymore and we hit an `EOFError`. Once we hit this error we know something is broken of course and there is a `_ssh_message_handler` method that checks for error messages and tries to raise a "nice" exception that is helpful and says what is wrong. If we don't see something we can match on, we will simply raise the strict key message you've seen. Logs should show us what was read, so hopefully this will indicate something that I haven't accounted for (and the "fix" may be just adding a cipher or kex or something) in the message parsing.

Thanks for raising this @991jo!

**Comment by carlmontanari** [comment]
Oh follow up -- I tested w/ my 2960 (Version 15.0(2)SE11,) and with my ssh config file having:

```
Host 172.31.254.4
  User vrnetlab
  Ciphers aes128-cbc
```

and a device setup like:

```
my_device = {
    "host": "172.31.254.4",
    "auth_username": "vrnetlab",
    "auth_password": "VR-netlab9",
    "auth_strict_key": False,
    "ssh_config_file": True
}
```

I am able to connect as expected.

If I modify the device to look like:

```
my_device = {
    "host": "172.31.254.4",
    "auth_username": "vrnetlab",
    "auth_password": "VR-netlab9",
    "auth_strict_key": False,
    "transport_options": {"open_cmd": ["-o", "Ciphers+=aes128-cbc"]},
}
```

and turn on logging, I get:

```
DEBUG:scrapli.transport-172.31.254.4:Session to host 172.31.254.4 spawned
DEBUG:scrapli.transport-172.31.254.4:Attempting to authenticate. Read: b'command-line: line 0: Bad configuration option: ciphers+\r\r\n'
CRITICAL:scrapli.transport-172.31.254.4:Failed to open connection to host 172.31.254.4. Do you need to disable `auth_strict_key`?
```

Changing the transport options from: ` {"open_cmd": ["-o", "Ciphers+=aes128-cbc"]},` to ` {"open_cmd": ["-o", "Ciphers=aes128-cbc"]},` (no "+=", just "=") makes things work as expected.

Can't say ive ever really ran into this as I pretty much always rely on ssh config file and drop the `+=` there... however this clearly is not appreciated by ssh as even manually running it as such (like scrapli basically does) breaks:

```
ssh 172.31.254.4 -p 22 -o ConnectTimeout=5 -o ServerAliveInterval=10 -l vrnetlab -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -F /dev/null -o Ciphers+=aes128-cbc
command-line: line 0: Bad configuration option: ciphers+
```

So thats just not going to work with scrapli or ssh in general is I think the end result here. Can you test things out on your end to validate what I've just seen? I'll remove the bug label I added since I guess this is hopefully not a bug!


Carl

**Comment by 991jo** [comment]
okay, I found my mistake, i have a typo in my code, the second `+`, is a `+=` not a `=+`.

```
wrong:
"transport_options": {"open_cmd": ["-o", "KexAlgorithms=+diffie-hellman-group-exchange-sha1", "-o", "Ciphers+=aes256-cbc"]},
right:
"transport_options": {"open_cmd": ["-o", "KexAlgorithms=+diffie-hellman-group-exchange-sha1", "-o", "Ciphers=+aes256-cbc"]},
```

now everything works as intended.

**Comment by carlmontanari** [comment]
Ahhh!! Nice catch I looked right at that and didn't catch it! I'll go ahead and close this then, glad it's working as expected!

**Comment by carlmontanari** [comment]
FYI I've improved some of the handling for the error messages for system ssh in develop -- should catch this issue in the future! Thanks again for opening this and working through it with me!
